tests/platform_helper/domain/test_codebase.py

```diff
@@ -0,0 +1,623 @@
+            ),
+        ]
+    )
+
+
+def test_codebase_build_does_not_trigger_build_without_an_application():
+    mocks = CodebaseMocks()
+    mocks.load_application_fn.side_effect = ApplicationNotFoundError()
+    codebase = Codebase(**mocks.params())
+
+    with pytest.raises(click.Abort) as exc:
+        codebase.build("not-an-application", "application", "ab1c23d")
+        mocks.echo_fn.assert_has_calls(
+            [
+                call(
+                    """The account "foo" does not contain the application "not-an-application"; ensure you have set the environment variable "AWS_PROFILE" correctly.""",
+                    fg="red",
+                ),
+            ]
+        )
+
+
+def test_codebase_prepare_does_not_generate_files_in_a_repo_with_a_copilot_directory(tmp_path):
+    mocks = CodebaseMocks()
+    mocks.load_application_fn.side_effect = SystemExit(1)
+    codebase = Codebase(**mocks.params())
+    os.chdir(tmp_path)
+    Path(tmp_path / "copilot").mkdir()
+
+    subprocess.run(["git", "init"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    subprocess.run(
+        ["git", "remote", "add", "origin", "git@github.com:uktrade/some-test-app.git"],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+
+    with pytest.raises(SystemExit) as exc:
+        codebase.prepare()
+
+    mocks.echo_fn.assert_has_calls(
+        [
+            call(
+                "You are in the deploy repository; make sure you are in the application codebase repository.",
+                fg="red",
+            ),
+        ]
+    )
+
+
+def test_codebase_prepare_generates_an_executable_image_build_run_file(tmp_path):
+    mocks = CodebaseMocks()
+    codebase = Codebase(**mocks.params())
+    os.chdir(tmp_path)
+
+    subprocess.run(["git", "init"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    subprocess.run(
+        ["git", "remote", "add", "origin", "git@github.com:uktrade/another-test-app.git"],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+
+    codebase.prepare()
+
+    image_build_run_path = Path(".copilot/image_build_run.sh")
+    assert image_build_run_path.exists()
+    assert image_build_run_path.is_file()
+    assert os.access(image_build_run_path, os.X_OK)
+
+
+@patch("subprocess.run")
+def test_codebase_build_does_not_trigger_build_without_confirmation(mock_subprocess_run):
+    mocks = CodebaseMocks()
+    mocks.confirm_fn.return_value = False
+    codebase = Codebase(**mocks.params())
+
+    mock_subprocess_run.return_value.stderr = ""
+
+    codebase.build("test-application", "application", "ab1c234")
+
+    mocks.echo_fn.assert_has_calls(
+        [
+            call(
+                """Your build was not triggered.""",
+            ),
+        ]
+    )
+
+
+def test_codebase_deploy_successfully_triggers_a_pipeline_based_deploy():
+    mocks = CodebaseMocks()
+    mocks.confirm_fn.return_value = True
+    mock_application = Application(name="test-application")
```

Comment:
is application mocking common, can it be extracted out in some way?

Reply:
done